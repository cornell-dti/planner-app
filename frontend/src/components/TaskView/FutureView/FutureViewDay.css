.NDaysView {
  position: relative;
  margin: -1em 4px 0 4px;
  background-color: #F7F7F7;
  height: 100%;
}

.OtherViews {
  position: relative;
  margin: -1px 0 0 -1px;
  background-color: white;
  border: 1px solid #CCC;
  height: 100%;
}

.NDaysView::-webkit-scrollbar, .OtherViews::-webkit-scrollbar {
  width: 0;
  height: 0;
}

.Today {
  background-color: white;
  box-shadow: 0 5px 12px 0 rgba(164, 164, 164, 0.3);
}

.DateInfo {
  margin: 1em;
  color: #3d3d3d;
  opacity: 0.57;
}

.DateInfoDay {
  font-weight: normal;
}

.DateNumNDaysView {
  font-weight: bold;
}

.DateNumOtherViews {
  font-weight: bold;
  font-size: 16px;
  color: #4A90E2;
}

.MoreTasksBar {
  position: absolute;
  width: 100%;
  bottom: 0;
  left: 0;
  right: 0;
  cursor: pointer;
  background-color: #ffffff;
  opacity: 1;
  color: #333333;
  border: 0;
  box-shadow: 0 -1px 1px rgba(0,0,0,0.1);
  text-align: center;
  line-height: 20px;
  font-size: 1em;
}

.FloatingViewPrevPadding {
  margin: 1em;
}

.FloatingView {
  position: absolute;
  z-index: 6;
  top: 0;
  left: 0;
  right: 0;
  max-height: 80vh;
  overflow-x: hidden;
  overflow-y: scroll;
  background-color: white;
  box-shadow: 0 5px 12px 0 rgba(164, 164, 164, 0.7);
}

.FloatingBackgroundBlocker {
  position: fixed;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  z-index: 5;
  background-color: rgba(0, 0, 0, 0.05);
}

.FloatingView::-webkit-scrollbar {
  width: 0;
  height: 0;
}

.FloatingView::-webkit-scrollbar-thumb {
  cursor: pointer;
  border-radius: unset;
  background-color: unset !important;
}
