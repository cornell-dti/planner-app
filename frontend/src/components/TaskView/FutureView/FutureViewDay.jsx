// @flow strict

import React from 'react';
import type { Node } from 'react';
import type { CompoundTask } from './future-view-types';
import type { FloatingPosition } from '../../Util/TaskEditors/task-editors-types';
import styles from './FutureViewDay.css';
import FutureViewDayTaskContainer from './FutureViewDayTaskContainer';
import {
  floatingViewWidth,
  nDaysViewHeaderHeight,
  otherViewsHeightHeader,
  taskContainerHeightNDaysView,
  taskContainerHeightOtherViews, taskHeight,
} from './future-view-css-props';
import { day2String } from '../../../util/datetime-util';
import windowSizeConnect from '../../Util/Responsive/WindowSizeConsumer';
import type { WindowSize } from '../../Util/Responsive/window-size-context';
import { error } from '../../../util/general-util';

type Props = {|
  +date: Date;
  +tasks: CompoundTask[];
  +inNDaysView: boolean;
  +taskEditorPosition: FloatingPosition;
  +windowSize: WindowSize;
|};

type State = {|
  +floatingViewOpened: boolean;
|};

type PropsForPositionComputation = {|
  +tasks: CompoundTask[];
  +inNDaysView: boolean;
  +windowSize: WindowSize;
  +mainViewPosition: {| +width: number; +height: number; +top: number; +left: number; |};
|};
type PositionStyle = {|
  +width: string;
  +height: string;
  +top: string;
  +left: string;
|};

/**
 * Returns the total number of tasks in the given task array.
 *
 * @param {CompoundTask[]} tasks the given task array.
 * @param {boolean} includeSubTasks whether to include subtasks.
 * @return {number} total number of tasks, including main task and subtasks.
 */
const countTasks = (tasks: CompoundTask[], includeSubTasks: boolean): number => {
  if (!includeSubTasks) {
    return tasks.length;
  }
  const subtaskReducer = v => v + 1;
  const reducer = (a: number, t: CompoundTask): number => (
    a + 1 + t.filtered.subtaskArray.reduce(subtaskReducer, 0)
  );
  return tasks.reduce(reducer, 0);
};

/**
 * Returns the computed floating view style from some properties.
 *
 * @return the computed style.
 */
const computeFloatingViewStyle = (props: PropsForPositionComputation): PositionStyle => {
  const {
    tasks, inNDaysView, windowSize,
    mainViewPosition: {
      width, height, top, left,
    },
  } = props;
  // Compute the height of inner content
  const headerHeight = inNDaysView ? nDaysViewHeaderHeight : otherViewsHeightHeader;
  const tasksHeight = taskHeight * countTasks(tasks, inNDaysView);
  const totalHeight = headerHeight + tasksHeight;
  // Decide the maximum allowed height and the actual height
  const maxAllowedHeight = inNDaysView ? 400 : 300;
  const floatingViewHeight = Math.min(totalHeight, maxAllowedHeight);
  // Compute ideal offset
  let topOffset = (height - floatingViewHeight) / 2;
  let leftOffset = (width - floatingViewWidth) / 2;
  // Correct the offsets if they overflow.
  {
    const windowWidth = windowSize.width;
    const windowHeight = windowSize.height;
    const topAbsolutePosition = top + topOffset;
    if (topAbsolutePosition < 0) {
      topOffset -= topAbsolutePosition;
    } else {
      const bottomAbsolutePosition = topAbsolutePosition + floatingViewHeight;
      const diff = bottomAbsolutePosition - windowHeight;
      if (diff > 0) {
        topOffset -= diff;
      }
    }
    const leftAbsolutePosition = left + leftOffset;
    if (leftAbsolutePosition < 0) {
      leftOffset -= leftAbsolutePosition;
    } else {
      const rightAbsolutePosition = leftAbsolutePosition + floatingViewWidth;
      const diff = rightAbsolutePosition - windowWidth;
      if (diff > 0) {
        leftOffset -= diff;
      }
    }
  }
  return {
    width: `${floatingViewWidth}px`,
    height: `${floatingViewHeight}px`,
    top: `${topOffset}px`,
    left: `${leftOffset}px`,
  };
};

/**
 * The component that renders all tasks on a certain day.
 */
class FutureViewDay extends React.PureComponent<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = { floatingViewOpened: false };
  }

  /**
   * Report whether the day is today.
   *
   * @return {boolean} whether the day is today.
   */
  isToday = (): boolean => {
    const { date } = this.props;
    const today = new Date();
    return date.getFullYear() === today.getFullYear()
      && date.getMonth() === today.getMonth()
      && date.getDate() === today.getDate();
  };

  /**
   * Report whether the container will overflow.
   *
   * @return {boolean} whether the container overflows.
   */
  doesOverFlow = (): boolean => {
    const { tasks, inNDaysView } = this.props;
    const totalRequiredHeight = taskHeight * countTasks(tasks, inNDaysView);
    const actualHeight = inNDaysView
      ? taskContainerHeightNDaysView
      : taskContainerHeightOtherViews;
    return totalRequiredHeight > actualHeight;
  };

  /**
   * Compute the floating view position.
   *
   * @return {PositionStyle} the floating view position.
   */
  computeFloatingViewPosition = (): PositionStyle => {
    const componentDiv = this.backlogDayElement ?? error('Impossible Case!');
    const boundingRect = componentDiv.getBoundingClientRect();
    if (!(boundingRect instanceof DOMRect)) {
      throw new Error('Bad boundingRect!');
    }
    const {
      width, height, top, left,
    } = boundingRect;
    const mainViewPosition = {
      width, height, top, left,
    };
    const { date, taskEditorPosition, ...positionProps } = this.props;
    return computeFloatingViewStyle({ ...positionProps, mainViewPosition });
  };

  /**
   * Open the floating view.
   */
  openFloatingView = () => this.setState({ floatingViewOpened: true });

  /**
   * Close the floating view.
   */
  closeFloatingView = () => this.setState({ floatingViewOpened: false });

  /**
   * Render the main content.
   *
   * @param {boolean} inMainList whether the header is in main list as opposed to floating list.
   * @return {*} the rendered header component.
   */
  renderContent = (inMainList: boolean): Node => {
    const {
      date, tasks, inNDaysView, taskEditorPosition,
    } = this.props;
    const isToday = this.isToday();
    const dateNumCssClass = inNDaysView
      ? styles.DateNumNDaysView
      : styles.DateNumOtherViews;
    const containerStyle = (inNDaysView && inMainList) ? { paddingTop: '1em' } : {};
    return (
      <React.Fragment>
        <div className={styles.DateInfo} style={containerStyle}>
          {inNDaysView && (
            <div className={styles.DateInfoDay}>
              {isToday ? 'TODAY' : day2String(date.getDay())}
            </div>
          )}
          <div className={dateNumCssClass}>{date.getDate()}</div>
        </div>
        <FutureViewDayTaskContainer
          tasks={tasks}
          inNDaysView={inNDaysView}
          taskEditorPosition={taskEditorPosition}
          isInMainList={inMainList}
        />
      </React.Fragment>
    );
  };

  /**
   * Render the body inside the day wrapper.
   *
   * @return {Node} the rendered body.
   */
  renderBody = (): Node => {
    if (!this.doesOverFlow()) {
      return this.renderContent(true);
    }
    const { floatingViewOpened } = this.state;
    if (!floatingViewOpened) {
      return (
        <React.Fragment>
          {this.renderContent(true)}
          <button type="button" className={styles.MoreTasksBar} onClick={this.openFloatingView}>
            More Tasks...
          </button>
        </React.Fragment>
      );
    }
    return (
      <React.Fragment>
        <div className={styles.FloatingViewPrevPadding} />
        <div
          role="presentation"
          className={styles.FloatingBackgroundBlocker}
          onClick={this.closeFloatingView}
        />
        <div className={styles.FloatingView} style={this.computeFloatingViewPosition()}>
          {this.renderContent(false)}
        </div>
      </React.Fragment>
    );
  };

  backlogDayElement: ?HTMLDivElement;

  render(): Node {
    const { inNDaysView } = this.props;
    let wrapperCssClass: string;
    if (inNDaysView) {
      wrapperCssClass = this.isToday()
        ? `${styles.NDaysView} ${styles.Today}` : styles.NDaysView;
    } else {
      wrapperCssClass = styles.OtherViews;
    }
    return (
      <div className={wrapperCssClass} ref={(e) => { this.backlogDayElement = e; }}>
        {this.renderBody()}
      </div>
    );
  }
}

const ConnectedFutureViewDay = windowSizeConnect<Props>(FutureViewDay);
export default ConnectedFutureViewDay;
