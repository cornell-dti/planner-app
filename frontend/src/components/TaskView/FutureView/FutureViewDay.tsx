import React, { ReactElement } from 'react';
import { SimpleDate } from './future-view-types';
import { CalendarPosition, FloatingPosition } from '../../Util/TaskEditors/editors-types';
import styles from './FutureViewDay.module.scss';
import { headerHeight } from './future-view-css-props';
import { getTodayAtZeroAM } from '../../../util/datetime-util';
import { error } from '../../../util/general-util';
import { useWindowSize, WindowSize } from '../../../hooks/window-size-hook';
import FutureViewDayContent from './FutureViewDayContent';

type Position = {
  readonly width: number;
  readonly height: number;
  readonly top: number;
  readonly left: number;
};
type PositionStyle = {
  readonly width: string;
  readonly height: string;
  readonly top: string;
  readonly left: string;
};
type PropsForPositionComputation = {
  readonly tasksHeight: number;
  readonly inNDaysView: boolean;
  readonly windowSize: WindowSize;
  readonly mainViewPosition: Position;
};

/**
 * Returns the computed floating view style from some properties.
 *
 * @return the computed style.
 */
const computeFloatingViewStyle = (props: PropsForPositionComputation): PositionStyle => {
  const {
    tasksHeight, inNDaysView, windowSize,
    mainViewPosition: { width, height, top },
  } = props;
  // Compute the height of inner content
  const totalHeight = headerHeight + tasksHeight;
  // Decide the maximum allowed height and the actual height
  const maxAllowedHeight = inNDaysView ? 500 : 300;
  const floatingViewHeight = Math.min(totalHeight, maxAllowedHeight);
  // Compute ideal offset
  let topOffset = (height - floatingViewHeight) / 2;
  // Correct the offsets if they overflow.
  {
    const windowHeight = windowSize.height;
    const topAbsolutePosition = top + topOffset;
    if (topAbsolutePosition < 0) {
      topOffset -= topAbsolutePosition;
    } else {
      const bottomAbsolutePosition = topAbsolutePosition + floatingViewHeight;
      const diff = bottomAbsolutePosition - windowHeight;
      if (diff > 0) {
        topOffset -= diff;
      }
    }
  }
  return {
    width: `${width}px`,
    height: `${floatingViewHeight}px`,
    top: `${topOffset}px`,
    left: '0',
  };
};

type Props = {
  readonly date: SimpleDate;
  readonly inNDaysView: boolean;
  readonly taskEditorPosition: FloatingPosition;
  readonly calendarPosition: CalendarPosition;
  readonly doesShowCompletedTasks: boolean;
};

type HeightInfo = { readonly doesOverflow: boolean; readonly tasksHeight: number };
const dummyHeightInfo: HeightInfo = { doesOverflow: false, tasksHeight: 0 };

/**
 * The component that renders all tasks on a certain day.
 */
export default function FutureViewDay(props: Props): ReactElement {
  const { date, inNDaysView, taskEditorPosition, calendarPosition, doesShowCompletedTasks } = props;
  const [floatingViewOpened, setFloatingViewOpened] = React.useState(false);
  const [heightInfo, setHeightInfo] = React.useState<HeightInfo>(dummyHeightInfo);
  const windowSize = useWindowSize();
  const componentDivRef = React.useRef<HTMLDivElement>(null);

  const onHeightChange = (doesOverflow: boolean, tasksHeight: number): void => {
    if (heightInfo.tasksHeight !== tasksHeight) {
      setHeightInfo({ doesOverflow, tasksHeight });
    }
  };

  const openFloatingView = (): void => setFloatingViewOpened(true);
  const closeFloatingView = (): void => setFloatingViewOpened(false);

  const isToday: boolean = getTodayAtZeroAM().toDateString() === date.text;
  let wrapperCssClass: string;
  if (inNDaysView) {
    wrapperCssClass = isToday ? `${styles.NDaysView} ${styles.Today}` : styles.NDaysView;
  } else {
    wrapperCssClass = isToday ? `${styles.OtherViews} ${styles.Today}` : styles.OtherViews;
  }
  if (!floatingViewOpened) {
    return (
      <div className={wrapperCssClass} ref={componentDivRef}>
        <FutureViewDayContent
          inMainList
          onHeightChange={onHeightChange}
          date={date}
          inNDaysView={inNDaysView}
          taskEditorPosition={taskEditorPosition}
          calendarPosition={calendarPosition}
          doesShowCompletedTasks={doesShowCompletedTasks}
        />
        {heightInfo.doesOverflow && (
          <button type="button" className={styles.MoreTasksBar} onClick={openFloatingView} tabIndex={0}>
            More Tasks...
          </button>
        )}
      </div>
    );
  }
  const computeFloatingViewPosition = (): PositionStyle => {
    const componentDiv = componentDivRef.current || error();
    const boundingRect = componentDiv.getBoundingClientRect();
    if (!(boundingRect instanceof DOMRect)) {
      throw new Error('Bad boundingRect!');
    }
    const {
      width, height, top, left,
    } = boundingRect;
    const mainViewPosition = {
      width, height, top, left,
    };
    const { tasksHeight } = heightInfo;
    return computeFloatingViewStyle({
      tasksHeight, inNDaysView, mainViewPosition, windowSize,
    });
  };
  return (
    <div className={wrapperCssClass} ref={componentDivRef}>
      <div className={styles.FloatingViewPrevPadding} />
      <div
        role="presentation"
        className={styles.FloatingBackgroundBlocker}
        onClick={closeFloatingView}
      />
      <div className={styles.FloatingView} style={computeFloatingViewPosition()}>
        <FutureViewDayContent
          inMainList={false}
          onHeightChange={onHeightChange}
          date={date}
          inNDaysView={inNDaysView}
          taskEditorPosition={taskEditorPosition}
          calendarPosition={calendarPosition}
          doesShowCompletedTasks={doesShowCompletedTasks}
        />
      </div>
    </div>
  );
}
