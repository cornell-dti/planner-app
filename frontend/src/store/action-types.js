// @flow strict

import type {
  PartialMainTask, PartialSubTask, SubTask, Tag, Task, Course,
} from './store-types';
import type { TaskDiff } from '../util/task-util';

/*
 * --------------------------------------------------------------------------------
 * Part 1: Task Actions
 * --------------------------------------------------------------------------------
 */

export type AddNewTaskAction = {| type: 'ADD_NEW_TASK'; +task: Task; |};
export type AddNewSubTaskAction = {|
  type: 'ADD_NEW_SUBTASK'; +taskId: number; +subTask: SubTask;
|};

export type EditTaskAction = {| type: 'EDIT_TASK'; +task: Task; +diff: TaskDiff; |};
export type EditMainTaskAction = {|
  type: 'EDIT_MAIN_TASK'; +taskId: number; +partialMainTask: PartialMainTask;
|};
export type EditSubTaskAction = {|
  type: 'EDIT_SUB_TASK'; +taskId: number; +subtaskId: number; +partialSubTask: PartialSubTask;
|};

export type RemoveTaskAction = {| type: 'REMOVE_TASK'; +taskId: number; |};
export type RemoveSubTaskAction = {|
  type: 'REMOVE_SUBTASK'; +taskId: number; +subtaskId: number;
|};

export type ImportCourseExamsAction = {| type: 'IMPORT_COURSE_EXAMS' |};

type TaskAction =
  | AddNewTaskAction
  | AddNewSubTaskAction
  | EditTaskAction
  | EditMainTaskAction
  | EditSubTaskAction
  | RemoveTaskAction
  | RemoveSubTaskAction
  | ImportCourseExamsAction;

/*
 * --------------------------------------------------------------------------------
 * Part 2: Color Config Actions
 * --------------------------------------------------------------------------------
 */

export type AddTagAction = {| type: 'ADD_TAG'; +tag: Tag; |};
export type EditTagAction = {| type: 'EDIT_TAG'; +tag: Tag; |};
export type RemoveTagAction = {| type: 'REMOVE_TAG'; +tagId: number; |};

export type TagAction = AddTagAction | EditTagAction | RemoveTagAction;

/*
 * --------------------------------------------------------------------------------
 * Part 3: Undo Actions
 * --------------------------------------------------------------------------------
 */

export type UndoAddTaskAction = {| type: 'UNDO_ADD_TASK' |};
export type ClearUndoAddTaskAction = {| type: 'CLEAR_UNDO_ADD_TASK' |};
export type UndoDeleteTaskAction = {| type: 'UNDO_DELETE_TASK' |};
export type ClearUndoDeleteTaskAction = {| type: 'CLEAR_UNDO_DELETE_TASK' |};

export type UndoAction =
  | UndoAddTaskAction
  | ClearUndoAddTaskAction
  | UndoDeleteTaskAction
  | ClearUndoDeleteTaskAction;

/*
 * --------------------------------------------------------------------------------
 * Part 4: Backend Patch Actions
 * --------------------------------------------------------------------------------
 */

export type BackendPatchNewTagAction = {|
  type: 'BACKEND_PATCH_NEW_TAG'; +tempId: number; +backendId: number;
|};
export type BackendPatchNewTaskAction = {|
  type: 'BACKEND_PATCH_NEW_TASK'; +tempId: number; +backendTask: Task;
|};
export type BackendPatchBatchNewTasksAction = {|
  type: 'BACKEND_PATCH_BATCH_NEW_TASKS'; +tempIds: number[]; +backendTasks: Task[];
|};
export type BackendPatchNewSubTaskAction = {|
  type: 'BACKEND_PATCH_NEW_SUBTASK';
  +taskId: number;
  +tempSubTaskId: number;
  +backendSubTaskId: number;
|};
export type BackendPatchExistingTaskAction = {|
  type: 'BACKEND_PATCH_EXISTING_TASK'; +task: Task;
|};
export type BackendPatchLoadedDataAction = {|
  type: 'BACKEND_PATCH_LOADED_DATA'; +tags: Tag[]; +tasks: Task[]; +courses: Map<number, Course[]>;
|};

/**
 * All the tasks that lets the backend patch the store with all the latest values.
 */
export type BackendPatchAction =
  | BackendPatchNewTagAction
  | BackendPatchNewTaskAction
  | BackendPatchBatchNewTasksAction
  | BackendPatchNewSubTaskAction
  | BackendPatchExistingTaskAction
  | BackendPatchLoadedDataAction

/*
 * --------------------------------------------------------------------------------
 * Part 5: All Actions
 * --------------------------------------------------------------------------------
 */

export type Action = TagAction | TaskAction | UndoAction | BackendPatchAction;
