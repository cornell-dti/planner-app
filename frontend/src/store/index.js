// @flow strict

import { createStore } from 'redux';
import type { Store } from 'redux';
import rootReducer from './reducers';
import type { State } from './store-types';
import type { Action } from './action-types';

const store: Store<State, $Subtype<Action>> = createStore(rootReducer);
export default store;
