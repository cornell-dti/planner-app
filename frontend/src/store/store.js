// @flow strict

import { createStore } from 'redux';
import type { Store } from 'redux';
import rootReducer from './reducers';
import type { State } from './store-types';
import type { Action } from './action-types';
import type { AppUser } from '../util/firebase-util';
import { error } from '../util/general-util';

export type GlobalStore = Store<State, $Subtype<Action>>;

const store: GlobalStore = createStore(rootReducer);
let appUser: AppUser | null = null;

/**
 * Initialize the app and record a global user and returns the initialized store.
 *
 * @param {AppUser} user the user of the app.
 * @return {GlobalStore} the newly initialized store.
 */
export function initializeApp(user: AppUser): GlobalStore {
  appUser = user;
  return store;
}

/**
 * Returns the global app user.
 *
 * @return {AppUser} the global app user.
 */
export const getAppUser = (): AppUser => appUser ?? error('App is not initialized.');

/**
 * Dispatch an action and returns the dispatched action.
 *
 * @param {Action} action the action to dispatch.
 * @return {Action} the action to dispatch.
 */
export const dispatchAction = (action: Action): Action => store.dispatch(action);
