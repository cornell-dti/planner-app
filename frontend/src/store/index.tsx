import React, { ReactElement } from 'react';
import { Provider as ReactReduxProvider } from 'react-redux';
import { store, storeForTesting } from './store';

type Props = { readonly children: ReactElement };

export const Provider = ({ children }: Props): ReactElement => (
  <ReactReduxProvider store={store}>{children}</ReactReduxProvider>
);

export const ProviderForTesting = ({ children }: Props): ReactElement => (
  <ReactReduxProvider store={storeForTesting}>{children}</ReactReduxProvider>
);
