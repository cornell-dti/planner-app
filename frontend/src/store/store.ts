import { createStore, Store } from 'redux';
import rootReducer from './reducers';
import { State } from './store-types';
import { Action } from './action-types';
import { initialStateForTesting } from './state';

export type GlobalStore = Store<State, Action>;

/**
 * The redux store. It should only be given to the react-redux provider.
 */
export const store: GlobalStore = createStore(rootReducer);

/**
 * The redux store for testing.
 */
export const storeForTesting: GlobalStore = createStore(
  (initialState: State = initialStateForTesting, action: Action): State => rootReducer(
    initialState, action,
  ),
);

/**
 * Dispatch an action and returns the same dispatched action.
 */
export const dispatchAction = (action: Action): Action => store.dispatch(action);
