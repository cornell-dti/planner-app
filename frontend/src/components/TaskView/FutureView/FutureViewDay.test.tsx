import React from 'react';
import renderer from 'react-test-renderer';
import { ProviderForTesting } from 'store';
import FutureViewDay from './FutureViewDay';

const createTest = (
  inNDaysView: boolean, doesShowCompletedTasks: boolean,
): void => {
  const testName = `FutureViewDay(${inNDaysView}, ${doesShowCompletedTasks}) matches snapshot.`;
  const date = new Date(Date.UTC(2030, 0, 1));
  const simpleDate = {
    year: date.getUTCFullYear(),
    month: date.getUTCMonth(),
    date: date.getUTCDate(),
    day: date.getUTCDay(),
    text: date.toDateString(),
  };
  it(testName, () => {
    const tree = renderer.create(
      <ProviderForTesting>
        <FutureViewDay
          date={simpleDate}
          inNDaysView={inNDaysView}
          taskEditorPosition="left"
          calendarPosition="top"
          doesShowCompletedTasks={doesShowCompletedTasks}
        />
      </ProviderForTesting>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
};
