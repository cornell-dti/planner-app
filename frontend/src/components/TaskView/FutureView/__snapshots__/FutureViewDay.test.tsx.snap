// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FutureViewDay(false, false) matches snapshot. 1`] = `
<div
  className="OtherViews"
>
  <div
    className="DateInfo"
    style={Object {}}
  >
    <div
      className="DateInfoDay"
    >
      TUE
    </div>
    <div
      className="DateNum"
    >
      1
    </div>
  </div>
  <div
    className="Container"
    style={
      Object {
        "overflow": "hidden",
      }
    }
  />
</div>
`;

exports[`FutureViewDay(false, false) matches snapshot. 2`] = `
<div
  className="OtherViews"
>
  <div
    className="DateInfo"
    style={Object {}}
  >
    <div
      className="DateInfoDay"
    >
      TUE
    </div>
    <div
      className="DateNum"
    >
      1
    </div>
  </div>
  <div
    className="Container"
    style={
      Object {
        "overflow": "hidden",
      }
    }
  />
</div>
`;

exports[`FutureViewDay(false, true) matches snapshot. 1`] = `
<div
  className="OtherViews"
>
  <div
    className="DateInfo"
    style={Object {}}
  >
    <div
      className="DateInfoDay"
    >
      TUE
    </div>
    <div
      className="DateNum"
    >
      1
    </div>
  </div>
  <div
    className="Container"
    style={
      Object {
        "overflow": "hidden",
      }
    }
  >
    <div
      className="Task"
      onClick={[Function]}
      onKeyUp={[Function]}
      role="button"
      style={Object {}}
      tabIndex={0}
    >
      <div
        className="TaskMainWrapper"
        style={
          Object {
            "backgroundColor": "red",
          }
        }
      >
        <label
          className="TaskCheckBox CheckBox"
        >
          <input
            defaultChecked={true}
            disabled={false}
            onClick={[Function]}
            tabIndex={0}
            type="checkbox"
          />
          <span
            className="CheckBoxCheckMark"
          />
        </label>
        <span
          className="TaskText"
          style={
            Object {
              "textDecoration": "line-through",
            }
          }
        >
          Foo
        </span>
        <span
          title="Pin to Focus"
        >
          <svg
            alt="Pin to Focus"
            className="TaskIcon"
            height="1em"
            onClick={[Function]}
            onKeyUp={[Function]}
            tabIndex={0}
            title="Pin to Focus"
            width="1em"
          >
            pin-2-light-outline.svg
          </svg>
        </span>
        <span
          title="Delete"
        >
          <svg
            alt="Delete"
            className="TaskIcon"
            height="1em"
            onClick={[Function]}
            onKeyUp={[Function]}
            tabIndex={0}
            title="Delete"
            width="1em"
          >
            XLight.svg
          </svg>
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`FutureViewDay(false, true) matches snapshot. 2`] = `
<div
  className="OtherViews"
>
  <div
    className="DateInfo"
    style={Object {}}
  >
    <div
      className="DateInfoDay"
    >
      TUE
    </div>
    <div
      className="DateNum"
    >
      1
    </div>
  </div>
  <div
    className="Container"
    style={
      Object {
        "overflow": "hidden",
      }
    }
  >
    <div
      className="Task"
      onClick={[Function]}
      onKeyUp={[Function]}
      role="button"
      style={Object {}}
      tabIndex={0}
    >
      <div
        className="TaskMainWrapper"
        style={
          Object {
            "backgroundColor": "red",
          }
        }
      >
        <label
          className="TaskCheckBox CheckBox"
        >
          <input
            defaultChecked={true}
            disabled={false}
            onClick={[Function]}
            tabIndex={0}
            type="checkbox"
          />
          <span
            className="CheckBoxCheckMark"
          />
        </label>
        <span
          className="TaskText"
          style={
            Object {
              "textDecoration": "line-through",
            }
          }
        >
          Foo
        </span>
        <span
          title="Pin to Focus"
        >
          <svg
            alt="Pin to Focus"
            className="TaskIcon"
            height="1em"
            onClick={[Function]}
            onKeyUp={[Function]}
            tabIndex={0}
            title="Pin to Focus"
            width="1em"
          >
            pin-2-light-outline.svg
          </svg>
        </span>
        <span
          title="Delete"
        >
          <svg
            alt="Delete"
            className="TaskIcon"
            height="1em"
            onClick={[Function]}
            onKeyUp={[Function]}
            tabIndex={0}
            title="Delete"
            width="1em"
          >
            XLight.svg
          </svg>
        </span>
      </div>
    </div>
  </div>
</div>
`;
